darwin.stdenv: improvements and overrideSDK rewrite

reckenrode opened the pull request:
## Description of changes

This PR makes several improvements to the Darwin stdenv situation.

### Adds SDK root

It adds the SDK version to the stdenv as a new `darwin.apple_sdk.sdkRoot` package. This is done via a hook that adds `-isysroot` with a path to a stub SDK that just defines version information (based on the `xcbuild.sdk`). This is needed because clang passes `-platform_version <deployment target> <sdk version>` unconditionally to the linker, and the SDK version is inferred to be the same as the deployment target if `-isysroot` or `SDKROOT` do not point to SDK from which it can get the version.
 
macOS implements different behaviors depending on the SDK version. There are two notable examples: dark mode and version checks.

macOS requires the linked SDK to be at least 10.14 for dark mode to work. Currently, x86_64-darwin binaries in nixpkgs do not work with dark mode. With the SDK root specified, clang will pass the correct SDK version, ensuring that binaries linked against the 11.0 SDK work as expected.

For version checks, macOS returns a compatible version when the linked SDK is older than 11.0. I ran into that issue while working on https://github.com/NixOS/nixpkgs/issues/236414. MetalD3D checks the macOS version, but because the load commands in Wine’s binary indicate an older SDK, macOS returns a “compatible” version: 10.20. Since the check is for 14, the check fails, and MetalD3D does not work as expected.

Closes #265139.

### Fleshes out `overrideSDK`

This PR includes a rewrite of `overrideSDK`, addressing most of the issues identified in https://github.com/NixOS/nixpkgs/issues/242666#issuecomment-1803866452. This was necessary to support building Rust applications such as lapce, neovide, and wezterm on x86_64-darwin using the 11.0 SDK. It is also necessary to fix building applications that use `wrapGAppsHook` and the 11.0 SDK.

According to ofborg’s evalulation performance check, the rewritten `overrideSDK` performs about the same as the previous version even though it is more capable: it supports replacing inputs of all types, supports additional SDK-based overrides (libobjc, libs, private frameworks), and should be easier to maintain. It should also work with cross-compilation, but this is untested due to needing https://github.com/NixOS/nixpkgs/pull/256590 to fix cross-compilation support.

I plan to address overriding the SDK for Rust in a separate PR. There are some changes to `rustPlatform` needed to make it work robustly with overriding the SDK. Documentation will also follow in a separate PR.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

reckenrode committed:
darwin.apple_sdk_10_12: add Libsystem

reckenrode committed:
overrideSDK: support all deps attributes

This is effectively a rewrite of `overrideSDK`. It was required because
`wrapGAppsHook` propagates `depsTargetTarget` with the expectation that
it will effectively be `buildInputs` when the hook is itself used as a
`nativeBuildInput`. This propagates Gtk, which itself propagates the
default Dariwn SDK, making it effectively impossible to override the SDK
when a package depends on Gtk and uses `wrapGAppsHook`.

This rewrite implements the following improvements:

* Cross-compilation should be supported correctly (untested);
* Supports public and private frameworks;
* Supports SDK `libs`;
* Remaps instead of replacing extra (native) build inputs in the stdenv;
* Updates any Darwin framework references in `nix-support`; and
* It updates `xcodebuild` regardless of which input its in.

The implementation avoids recursion for performance reasons. Instead, it
enumerates transitive dependencies and walks the list from the leaf
packages backwards to the parent packages.

reckenrode committed:
darwin.apple_sdk.sdkRoot: init at 10.12 and 11.0

Based on the derivation from xcbuild.sdk. apple_sdk.sdkRoot provides
version plists and a hook that passes them automatically to the compiler
as `-isysroot`. This is needed to correctly set the SDK version in
compiled binaries. Otherwise, clang will infer the SDK version to be the
same as the deployment target, which is usually not what you want.

reckenrode committed:
darwin.stdenv: add sdkRoot to extraNativeBuildInputs

Setting the SDK root by default allows `overrideSDK` to correctly set
the SDK version when using a different SDK. It also allows the correct
SDK version to be set when using an older deployment target. Not setting
the correct SDK version can result in unexpected behavior at runtime.

Examples:

* Automatic dark mode switching requires linking against an SDK version
  of 10.14 or newer. With the current behavior, the only way to do this
  is by using a 10.14+ deployment target even when the application
  supports older platforms when build with a newer SDK.
* MetalD3D checks that the system version is at least 14.0. The API it
  uses returns a compatibility version when the the SDK is older than
  11.0, which causes it to display an error and terminate the
  application even when even when its requirements are all met.

reckenrode committed:
darwin.apple_sdk_10_12: add objc4

reckenrode commented:
@toonn

reckenrode commented:
The force push added support for multiple outputs in the override (though I don’t believe any of the packages currently overridden require it).

toonn reviewed (commented):
Everything looks good so far. I'll look at override-sdk.nix in another review. Didn't want to have you wait on the rest.

reckenrode commented on pkgs/os-specific/darwin/apple-sdk/sdkRoot.nix:
```diff
@@ -0,0 +1,66 @@
+ #
+ # See comments in cc-wrapper's setup hook. This works exactly the same way.
+ #
+ # TODO(@Ericson2314): No native exception
```
I wasn’t sure either. I copied it from cc-wrapper. I can drop its

reckenrode replied:
Dropped. I pushed an update that reforms `override-sdk.nix` according to the new RFC and adds support for compilers to the mapping. If it has a libc, the libc will get swapped to the requested SDK’s libc.

reckenrode commented:
Aside from adding compiler support, I added an optimization to reduce evaluation time by only building the mapping once per unique package set instead of doing it for every one then deduping at the end. I’m curious about the ofborg performance evaluation because the previous one was *terrible*.

reckenrode commented:
I managed to improve the terrible performance by ~30%, but the increase in evaluation time is still terrible. Need to look into some stuff on optimizing nix expressions without losing the generality of this solution.

reckenrode commented:
I fixed the merge conflict. I also optimized the algorithm. I left it as a separate commit, so the non-optimized version can serve as a reference. Running the ofborg check locally, it’s only a 4.6% regression (from 386s to 404s). That’s a _lot_ better considering it was as high as 600% and was 400% after my last attempt at optimization.

What I ended up doing was using `genericClosure` to calculate all dependencies for an input, then I sort and fold over them to calculate the replacements. I then update the original inputs using the new mapping I generate from that.

I’m aware of an issue with Qt 6 where it gets the wrong SDK root. The other overrides appear to work correctly though, so that’s strange. I’m planning to look into that next.

reckenrode commented:
The Qt 6 issue is because it’s using `apple_sdk_11_0.stdenv` instead of using `overrideSDK`.

reckenrode commented:
@ofborg eval

toonn reviewed (commented):
Mostly noticed some typos.

I honestly can't say I've understood everything in detail and I haven't done any testing yet but I wanted to provide at least a first review of overrideSDK.

I wish it was possible to with a single genericClosure invocation rather than the toposort but I think that would require the alternative solution you had in mind with memoization and that does sound worse than the current implementation.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,412 @@
+      inherit (newVersion) darwinMinVersion darwinSdkVersion;
+
+      # Used to get an SDK version corresponding to the requested `darwinSdkVersion`.
+      # TODO: Treat `darwinSdkVersion` as a constraint rather than as an exact version.
```
Treating it as a constraint is needed to make updating SDKs not be a breaking change. Otherwise, updating from the 14.0 to the 14.3 SDK would break everything that used `overrideSDK stdenv "14.0";`. The constraint for a bare number would be similar to a `^` constraint with some special handling for Darwin SDK versioning.

For example:`14.0` would match `14.3` but not `15`; `10.13` would match `10.13.6` but not `10.14`.

(Supporting other types of constraints later is possible, but I don’t see a use case for SDK versioning.)

toonn replied:
The interpretation of a string argument can change, that seems sensible. But I could see a use for `darwinMaxSdkVersion` for example, probably with a name that doesn't imply it means something similar to `darwinMinSdkVersion`.

It's true that the APIs are probably stable enough that it's not very important though.

reckenrode replied:
That could be done as a new constraint (such as `<=14` for the newest SDK prior to the 14.0 SDK). I just don’t want to support it without an actual need for it yet.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,412 @@
+        lib.pipe sdkPackages [
+          (lib.filter (attrs: attrs ? replacement))
+          (map (
+            { key, replacement }:
```
The sets here only have two keys: `key` and one of `pkgs, `replacement`, or `replacementList`. It’s not an issue because the filter should remove those other two. If one of those creeps in, it’s an error, and evaluation should fail. The only reason for the remapping is make everything conform to `listToAttrs`.

One possibility is defining `name` as an alias for `key` and rename `replacement` to `value`. That should let me drop the remapping step. Would that be better?

toonn replied:
No, the current way is fine. I just wasn't certain the attributes were exclusive.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,412 @@
+                targetFile="''${!outputName}/nix-support/$fileName"
+
+                sed -f "$dependenciesPath" < "$file" > "$targetFile"
+                if cmp -s "$file" "$targetFile"; then
```
I could do a prepass with `grep`. I was trying to keep things simple.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,412 @@
+        getPropagatedDependencies
+        (lib.toposort (x: y: lib.any (z: x.key == z) y.deps))
+        (lib.getAttr "result")
+        lib.reverseList
```
I don’t think that will work. `toposort` is doing a DFS using the comparison function to define which nodes are adjacent to the current node. I’d need to change `getPropagatedDependencies` to return a mapping of dependency to parents, which doesn’t seem worth doing just to avoid a `reverseList`.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,412 @@
+        newPackages
+      else
+        newPackages // { ${outPath} = replacement; }
+    ) baseMapping input;
```
That’s … interesting. I wonder how this is even working or if it’s working. It’s working on the input’s direct dependencies, which is almost certainly wrong. Hmm.

reckenrode replied:
It took some work, but I was able to make it use `dependencies`. I had to deal with the fact that inputs are not always derivations.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,412 @@
+      # outputs and the contents of `nix-support`, which will be customized.
+      ''
+        for outputName in $outputs; do
+          mkdir -p "''${!outputName}"
```
If I don’t create the output, won’t the build fail?

reckenrode replied:
I see what you mean now. Yeah, that’s a problem. The original iterated over the source paths, but I dropped that to avoid having to pass them as a file and parse them back out. The same reasoning is what motivated me to try to use sed to do the replacements. I’ll look into a solution for both.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,412 @@
+              mkdir "''${!outputName}/nix-support"
+
+              for file in "$targetPath"/*; do
+                fileName=$(basename "$file")
+                targetFile="''${!outputName}/nix-support/$fileName"
```
Good catch. It is not guaranteed. ruby-modules puts files in `nix-support/gem-meta`.

reckenrode commented:
@toonn I think I addressed most of the feedback in the latest commit. I’m refraining from squashing due to the complexity of the changes. Once everything is approved, I can squash the optimization commit down.

The only thing I haven’t addressed is the remapping logic. I need to recursively walk `nix-support` and figure out a solution to the replacement issue.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,412 @@
+            let
+              replacement = getReplacement newPackages dep;
+            in
+            lib.optionalString (dep != replacement) "s|${dep}|${replacement}|g;"
```
Would `lib.escapeRegex` be sufficient to avoid that problem?

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,412 @@
+    else
+      pkg;
+
+  propagatedInputs = [
```
I used it in an earlier version of `getPropagatedInputs`, but it’s not needed separately anymore. I’ve removed it from the current version.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,412 @@
+          key = dep.outPath;
+          deps = getPropagatedInputs dep;
+        }) deps;
+    };
```
I ended up refactoring this to reuse the function since it’s effectively the same.

reckenrode committed:
darwin.stdenv: use wrapBintoolsWith instead of a direct import

reckenrode commented:
Setting to draft for now while I figure out why Qt 5 is getting mixed SDK dependencies.

reckenrode commented:
The Qt 5 issue appears to be that the package built by luck before. It is getting both 10.12 and 11.0 SDK frameworks on master, but the 11.0 SDK frameworks were listed first in `NIX_CFLAGS_COMPILE`, so clang would find 11.0 SDK headers before 10.12 SDK headers. It was using the 10.12 Libsystem, which I’m surprised worked.

toonn reviewed (commented):
Figured it's more convenient to have the review sooner rather than waiting for you to squash and then you having to wait for the review.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,417 @@
+    lib.concatMap expandOutputs [
+      # Libsystem needs to match the one used by the SDK or weird errors happen.
+      {
+        original = pkgs.darwin.apple_sdk.Libsystem;
```
Remapping is done by `outPath`. `darwin.Libsystem` and `apple_sdk.Libsystem` have the same paths.

toonn replied:
Even on current x86_64-darwin?

reckenrode replied:
Yes. I set it in 36202add013a8a81b75b61de4f53955fd91b2900.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,417 @@
+        mappings
+    ) [ ] pkgs.darwin.apple_sdk.${attr};
+
+  # Produces a list of overrides for the given package set and SDK Version.
```
`version` is a set. It’s a workaround for dealing with xcodebuild. It needs some work not to hardcode its SDK root and toolchain stuff, but that’s out of scope for this PR. Once that’s done, `version` and the xcodebuild override can be dropped because it should do the right thing regardless of which SDK it was built against.

toonn replied:
And we can't extract the version set from the sdkRoot package?

reckenrode replied:
It’s more that I don’t want to make this bigger than it already is by also making changes to xcodebuild. xcodebuild needs to set the SDK root, but it also creates a toolchain from the clang in its stdenv.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,417 @@
+          coreutils = bintools.coreutils_bin;
+          gnugrep = bintools.gnugrep_bin;
+
+          inherit (bintools) bintools;
```
It shouldn’t refer to the original wrapper. `bintools.bintools` is the unwrapped bintools. `bintools.coreutils_bin` and bintools.gnugrep_bin` are the coreutils and gnugrep packages that were passed into the previous, wrapped bintools. I have to rewrap because I can’t use `override` to replace the libc.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,417 @@
+  #
+  # Note: `env` is an attrset containing `outputs` and `dependencies`.
+  # `dependencies` is a regex passed to sed and must be `passAsFile`.
+  mkStub =
```
Considering I refer to it as a proxy package elsewhere, I’ll update this to `mkProxyPackage` for consistency.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,417 @@
+          targetPath=$2
+          targetName=$(basename "$targetPath")
+
+          mkdir "$targetPath"
```
It should only ever recurse one level at a time. `$targetPath` will either be `${!outputName}/nix-support` or a subfolder of that.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,417 @@
+          in
+          lib.optionalString (
+            dep != replacement
+          ) "s|${lib.escapeRegex dep.outPath}|${lib.escapeRegex replacement.outPath}|g;"
```
I’m going to look at other solutions. I might do the replacements with `substitute`, and hopefully the performance won’t be too bad.

reckenrode replied:
The solution I am pursuing is setting `__structuredAttrs = true;`, which will serialized the attributes as JSON and set them as associative arrays in bash. That lets me avoid parsing the values using `passAsFile`. For replacements, I’m just using `substituteInPlace` iteratively over the `replacements` associative array.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,417 @@
+      pkg: newPackages:
+      let
+        replacement = replacePropagatedPackages newPackages pkg.package;
+        outPath = builtins.unsafeDiscardStringContext pkg.key;
```
Yeah, that’s redundant. I’ll drop it.

reckenrode commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,417 @@
+    }) mapping.original.outputs;
+
+  # Produces a list of mappings from an SDK path specified by `attr` in the given
+  # package set and SDK version.
```
I attempted to reword the comment to better explain what’s happening.

toonn commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,437 @@
+
+      inherit (env) outputs replacements sourceOutputs;
+
+      # Take advantage of the fact that replacements and sourceOutPaths will be passed
```
```suggestion
      # Take advantage of the fact that replacements and sourceOutputs will be passed
```

reckenrode replied:
Good catch. I originally called the variable `sourceOutPaths` but switched it because `sourceOutputs` read better in the build script.

toonn commented on pkgs/stdenv/darwin/override-sdk.nix:
```diff
@@ -0,0 +1,437 @@
+              if cmp -s "$sourceFile" "$targetFile"; then
+                rm "$targetFile"
+                ln -s "$sourceFile" "$targetFile"
+              fi
```
Would it be worth keeping track of whether any file ever differs from the sourceFile here? In the event that happens we can replace the entire package with a single symlink to the source package.

reckenrode replied:
`mkProxyPackage` checks whether `replacements` is empty before building the package (and returns the original package if it is). If it’s non-empty, that means there’s at least one input that will be replaces in `nix-support`.

When the `replacePropagatedInputs` bash function recurses, it tries to replace the folder with a symlink if it contains only symlinks. That is untested due to a lack of packages with propagated inputs in a subfolder.

toonn replied:
Ah, so you're checking if the replacements applied to a file, not whether the replacements are actually not a change at all?

reckenrode replied:
When `replacements` is generated in nix, any replacements that match the original are filtered out. If that results in an empty set, then the build is skipped.

By the time the build is being run, it should find at least one replacement. If it doesn’t find any, that would mean a package has propagated inputs that are not recorded in (at least) `$out/nix-support/*propagated*`.

I don’t know how that would happen because recording propagated inputs done automatically for you. The package would have to do something strange on its own to subvert that.

I could try to detect that situation and turn the package into a symlink back to the original, but I hesitate to make that change without having a test case.

toonn reviewed (commented):
One more typo and one more question but IMO this is pretty much done. :+1:

toonn reviewed (approved):
LGTM

reckenrode commented:
@AndersonTorres Sorry about the ping. A Meson I need to build locally patch accidentally snuck into the PR.

reckenrode commented:
The latest force push rebases on current staging to pick up the libiconv changes.

reckenrode commented:
@toonn I squashed the wip commits and set the status to ready. I also added a commit dropping unused curl assertions from the stdenv bootstrap. The contents of the PR should otherwise be the same as before the squash.

reckenrode committed:
darwin.stdenv: drop curl from assertions

toonn reviewed (approved):
Still LGTM : )

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/darwin-updates-news/42249/7
